crypto/src/immutable_bytes.rs

```diff
@@ -0,0 +1,90 @@
+#[cfg(not(feature = "std"))]
+use alloc::{boxed::Box, format, string::String, vec, vec::Vec};
+use core::mem;
+#[cfg(feature = "std")]
+use std::{boxed::Box, vec::Vec};
+
+use iroha_schema::IntoSchema;
+use parity_scale_codec::{Compact, Decode, Encode};
+use serde::{Deserialize, Serialize};
+
+/// Stores bytes that are not supposed to change during the runtime of the program in a compact way
+///
+/// This is a more efficient than `Vec<u8>` because it does not have to store the capacity field
+#[derive(
+    Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default, Serialize, Deserialize, IntoSchema,
+)]
+#[schema(transparent = "Vec<u8>")]
+pub struct CompactBytes(Box<[u8]>);
+
+impl CompactBytes {
+    pub fn new(bytes: impl Into<Box<[u8]>>) -> Self {
+        Self(bytes.into())
+    }
+    pub fn from_hex<T: AsRef<[u8]> + ?Sized>(payload: &T) -> Result<Self, crate::Error> {
+        Ok(Self::new(crate::hex_decode(payload)?))
+    }
+
+    /// Converts the `ImmutableBytes` into a `Vec<u8>`, reusing the heap allocation.
+    pub fn into_vec(self) -> Vec<u8> {
+        self.0.into_vec()
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+impl AsRef<[u8]> for CompactBytes {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+// Deriving Encode/Decode doesn't work for some reason, so we implement it manually.
+// The encoded representation is the same as &[u8] or Vec<u8>
+impl Encode for CompactBytes {
+    fn size_hint(&self) -> usize {
+        mem::size_of::<u32>() + self.len()
+    }
+
+    fn encode_to<T: parity_scale_codec::Output + ?Sized>(&self, dest: &mut T) {
+        let len = u32::try_from(self.len())
+            .expect("Attempted to serialize an ImmutableBytes that is too large.");
+        let len = Compact(len);
+        <Compact<u32>>::encode_to(&len, dest);
+
+        dest.write(self.as_ref());
+    }
+}
+
+impl Decode for CompactBytes {
+    fn decode<I: parity_scale_codec::Input>(
+        input: &mut I,
+    ) -> Result<Self, parity_scale_codec::Error> {
+        let Compact(length) = <Compact<u32>>::decode(input)?;
+        let mut out = vec![0u8; length as usize];
+        input.read(&mut out)?;
+        Ok(Self::new(out))
+    }
+}
```

Comment:
I think that if you don't specify in the schema that you're using compact encoding it will break JDKs. Actually, if you look at `ConstString`, I think you can delegate encoding to `Vec<u8>` like this:

```rs
impl WrapperTypeEncode for CompactBytes {}

impl WrapperTypeDecode for CompactBytes {
    type Wrapped = Vec<u8>;
}
```

Reply:
Didn't notice that such a mechanism existed. Using `WrapperTypeEncode`/`WrapperTypeDecode` now